### PR TITLE
kvserver: add logging if a lease {request, transfer} is not applied 

### PR DESCRIPTION
**Pull request by arulajmani**
See individual commits for details.

### CODE REVIEW

File: pkg/kv/kvserver/kvserverbase/forced_error.go [hunk 1]
```diff
@@ -72,34 +72,51 @@ func CheckForcedErr(
...
+		requestedLease = *raftCmd.ReplicatedEvalResult.State.Lease
+	}
+
+	defer func() {
+		if res.ForcedError != nil && isLeaseRequest {
+			log.Infof(ctx,
+				"rejected lease request %s; current lease %s; err: %s",
+				requestedLease, *replicaState.Lease, res.ForcedError)
+		}
+		if raftCmd.ReplicatedEvalResult.IsLeaseTransfer {
```
Comment (stevendanna) [inline pkg/kv/kvserver/kvserverbase/forced_error.go:88]: It looks like this isn't conditional on a non-nil error but was intended to be.  Perhaps we can early return for a nil error at the beginning of the defer?

Comment (tbg) [reply pkg/kv/kvserver/kvserverbase/forced_error.go:88]: Yep can confirm that this logs spuriously.

> I241114 11:46:35.636960 2226 kv/kvserver/kvserverbase/forced_error.go:90 ⋮ [T1,Vsystem,n3,s3,r3540/3:‹/Table/106/1/-418{4504…-3889…}›,raft] 14476  rejected lease transfer repl=(n2,s2):2 seq=7 start=1731584795.531805573,0 exp=1731584801.528864866,0 pro=1731584795.528864866,0 acq=Transfer; current lease repl=(n3,s3):3 seq=6 start=1731582944.484704600,0 epo=4 min-exp=0,0 pro=1731582944.484704600,0 acq=Request; err: <nil>

Comment (arulajmani) [reply pkg/kv/kvserver/kvserverbase/forced_error.go]: Yup, it was. Thanks for catching -- fixed.

**Review by arulajmani — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135129)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @stevendanna)

**Review by nvb — COMMENTED** [review]
> Reviewed 1 of 1 files at r1, 3 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135129)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @stevendanna, and @tbg)

___
*[`pkg/kv/kvserver/kvserverbase/forced_error.go` line 89 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/135129#-OBfvwqJ8CntoEe2yZa4:-OBfvwqJ8CntoEe2yZa5:bx8bf3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc4d8fb459c732da4386acef3a5bf46e327dbcb/pkg/kv/kvserver/kvserverbase/forced_error.go#L89)):*
> ```Go
> 		}
> 		if res.ForcedError != nil && raftCmd.ReplicatedEvalResult.IsLeaseTransfer {
> 			transferredLease := *raftCmd.ReplicatedEvalResult.State.Lease
> ```

Why not just use `requestedLease`?

In fact, much of this logic could be merged with the logic above. I'm thinking something along the lines of:
```go
defer func() {
	if res.ForcedError != nil && requestedLease != nil {
		op := redact.SafeString("transfer")
		if isLeaseRequest {
			op = "request"
		}
		log.Infof(ctx,
			"rejected lease %s %s; current lease %s; err: %s",
			op, *requestedLease, *replicaState.Lease, res.ForcedError)
	}
}()
```
___
*[`pkg/kv/kvserver/kvserverpb/proposer_kv.proto` line 121 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/135129#-OBfvIrcETzSiyP_Bsyh:-OBfvIrcETzSiyP_Bsyi:b-8fi39y) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc4d8fb459c732da4386acef3a5bf46e327dbcb/pkg/kv/kvserver/kvserverpb/proposer_kv.proto#L121)):*
> ```proto
>   ComputeChecksum compute_checksum = 21;
>   bool is_lease_request = 6;
>   bool is_lease_transfer = 28;
> ```

I'd strongly suggest that we don't add to a replicated proto that we're going to push through raft if we don't need to. Once we make these kinds of changes, they're very hard to undo, so they end up sitting around for years.

Can't we detect this in `CheckedForcedError` through the condition `raftCmd.ReplicatedEvalResult.State.Lease != nil && !raftCmd.ReplicatedEvalResult.IsLeaseRequest`?
___
*[`pkg/kv/kvserver/kvserverbase/forced_error.go` line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135129#-OBfuRv18eFaAzBS-JoT:-OBfuRv18eFaAzBS-JoU:bm7b27m) ([raw file](https://github.com/cockroachdb/cockroach/blob/19246ab45db33b9d39ecd20a2260f3481163cff5/pkg/kv/kvserver/kvserverbase/forced_error.go#L77)):*
> ```Go
> ) (res ForcedErrResult) {
> 	isLeaseRequest := raftCmd.ReplicatedEvalResult.IsLeaseRequest
> 	var requestedLease roachpb.Lease
> ```

I'm concerned about this escaping to the heap. One easy way to avoid that is to make it a pointer. So
```go
var requestedLease *roachpb.Lease
if isLeaseRequest {
	requestedLease = raftCmd.ReplicatedEvalResult.State.Lease
}
```
___
*[`pkg/kv/kvserver/kvserverbase/forced_error.go` line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135129#-OBfuoRw-0XEILyd1Prf:-OBfuoRw-0XEILyd1Prg:b-5i7sis) ([raw file](https://github.com/cockroachdb/cockroach/blob/19246ab45db33b9d39ecd20a2260f3481163cff5/pkg/kv/kvserver/kvserverbase/forced_error.go#L86)):*
> ```Go
> 			log.Infof(ctx,
> 				"rejected lease request %s; current lease %s; err: %s",
> 				requestedLease, *replicaState.Lease, res.ForcedError)
> ```

nit: does this logging still look correct if we don't explicitly dereference the `*roachpb.Lease` values?
___
*[`pkg/kv/kvserver/kvserverbase/forced_error.go` line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135129#-OBfv1H9779viJ3pytrV:-OBfv1H9779viJ3pytrW:bqj2zko) ([raw file](https://github.com/cockroachdb/cockroach/blob/19246ab45db33b9d39ecd20a2260f3481163cff5/pkg/kv/kvserver/kvserverbase/forced_error.go#L94)):*
> ```Go
> 		// that we can special case as indicating success of the probe above
> 		// raft.
> 		res = ForcedErrResult{
> ```

These `res` assignments shouldn't be needed for the defer to work correctly. Were you finding that it wasn't working without them?

**Review by arulajmani — COMMENTED** [review]
Sanity checked these log lines look as expected. Should be good for another look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135129#-:-OBg5nlU0fcTqIORAZr3:b-8moug1)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @stevendanna, and @tbg)

___
*[`pkg/kv/kvserver/kvserverbase/forced_error.go` line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135129#-OBfuRv18eFaAzBS-JoT:-OBg4MlO8go2PyFnCR1j:bmp5ieq) ([raw file](https://github.com/cockroachdb/cockroach/blob/19246ab45db33b9d39ecd20a2260f3481163cff5/pkg/kv/kvserver/kvserverbase/forced_error.go#L77)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm concerned about this escaping to the heap. One easy way to avoid that is to make it a pointer. So
```go
var requestedLease *roachpb.Lease
if isLeaseRequest {
	requestedLease = raftCmd.ReplicatedEvalResult.State.Lease
}
```
</blockquote></details>

I reverted this bit of the change, now that we're not using it in the deferred function. 
___
*[`pkg/kv/kvserver/kvserverbase/forced_error.go` line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135129#-OBfuoRw-0XEILyd1Prf:-OBg4L2H3fosuJAUrrVF:bxf6dqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/19246ab45db33b9d39ecd20a2260f3481163cff5/pkg/kv/kvserver/kvserverbase/forced_error.go#L86)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: does this logging still look correct if we don't explicitly dereference the `*roachpb.Lease` values?
</blockquote></details>

It does. Fixed.
___
*[`pkg/kv/kvserver/kvserverbase/forced_error.go` line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135129#-OBfv1H9779viJ3pytrV:-OBg4Fy33cfkVMH8OX5h:b-65vq8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/19246ab45db33b9d39ecd20a2260f3481163cff5/pkg/kv/kvserver/kvserverbase/forced_error.go#L94)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

These `res` assignments shouldn't be needed for the defer to work correctly. Were you finding that it wasn't working without them?
</blockquote></details>

I thought we needed it, but turns out we don't. Fixed. 
___
*[`pkg/kv/kvserver/kvserverpb/proposer_kv.proto` line 121 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/135129#-OBfvIrcETzSiyP_Bsyh:-OBg4kDA0Kt2r1ajBWO3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc4d8fb459c732da4386acef3a5bf46e327dbcb/pkg/kv/kvserver/kvserverpb/proposer_kv.proto#L121)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'd strongly suggest that we don't add to a replicated proto that we're going to push through raft if we don't need to. Once we make these kinds of changes, they're very hard to undo, so they end up sitting around for years.

Can't we detect this in `CheckedForcedError` through the condition `raftCmd.ReplicatedEvalResult.State.Lease != nil && !raftCmd.ReplicatedEvalResult.IsLeaseRequest`?
</blockquote></details>

Done.

**Review by nvb — APPROVED** [review]
:lgtm:

> Reviewed 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135129#-:-OBgA5Wg8dchSJQtOb_P:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @stevendanna, and @tbg)

___
*[`pkg/kv/kvserver/kvserverbase/forced_error.go` line 12 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/135129#-OBg9cd68257Xx1s7JTb:-OBg9cd68257Xx1s7JTc:b-3ez25z) ([raw file](https://github.com/cockroachdb/cockroach/blob/33c258032cf737e7386f02d669129bb4f712cbe6/pkg/kv/kvserver/kvserverbase/forced_error.go#L12)):*
> ```Go
> 	"context"
> 	"fmt"
> 	"github.com/cockroachdb/redact"
> ```

Lint probably won't like this.

**Comment by arulajmani** [comment]
Flake, I think. 

bors r+

**Comment by tbg** [comment]
```
  --- FAIL: acceptance/mismatched-locality (424.07s)
  (mismatched_locality.go:58).runMismatchedLocalityTest: dial tcp 127.0.0.1:29114: connect: connection refused
  test artifacts and logs in: /github-actions-runner/_work/cockroach/cockroach/artifacts/acceptance/mismatched-locality/run_1
  2024/11/15 02:18:03 test_runner.go:1194: ::error title=acceptance/mismatched-locality failed::(mismatched_locality.go:58).runMismatchedLocalityTest: dial tcp 127.0.0.1:29114: connect: connection refused
  Error: (mismatched_locality.go:58).runMismatchedLocalityTest: dial tcp 127.0.0.1:29114: connect: connection refused
  2024/11/15 02:18:03 test_runner.go:1194: ::error title=acceptance/mismatched-locality failed::test artifacts and logs in: /github-actions-runner/_work/cockroach/cockroach/artifacts/acceptance/mismatched-locality/run_1
```

I doubt this is this PR.

bors r+

**Comment by tbg** [comment]
> ```
>   --- FAIL: acceptance/mismatched-locality (424.07s)
>   (mismatched_locality.go:58).runMismatchedLocalityTest: dial tcp 127.0.0.1:29114: connect: connection refused
>   test artifacts and logs in: /github-actions-runner/_work/cockroach/cockroach/artifacts/acceptance/mismatched-locality/run_1
>   2024/11/15 02:18:03 test_runner.go:1194: ::error title=acceptance/mismatched-locality failed::(mismatched_locality.go:58).runMismatchedLocalityTest: dial tcp 127.0.0.1:29114: connect: connection refused
>   Error: (mismatched_locality.go:58).runMismatchedLocalityTest: dial tcp 127.0.0.1:29114: connect: connection refused
>   2024/11/15 02:18:03 test_runner.go:1194: ::error title=acceptance/mismatched-locality failed::test artifacts and logs in: /github-actions-runner/_work/cockroach/cockroach/artifacts/acceptance/mismatched-locality/run_1
> ```
> 
> I doubt this is this PR. Started a thread [here](https://cockroachlabs.slack.com/archives/CJ0H8Q97C/p1731672880726929) to take a look at the artifacts. bors r+

It was unrelated: https://github.com/cockroachdb/cockroach/issues/135273

**Pull request merged**